syntax = "proto3";
package Protocol;

    // Type
    enum ModuleType{
        Unconfigured = 0;
        Control = 1;
        Firewall = 2;
        Gateway = 3;
        Data = 4;
        Process = 5;
        DB = 6;
    }

    enum MsgTypes{
        Request = 0;
        Response = 1;
    }

message MessagePackage
{
    // 功能
    int32 Function = 1;
    // 优先级
    int32 PriorityLevel = 2;
    // 流水号
    int32 SerialNumber = 3;
    // 校验串
    string CheckString = 4;
    // 发送者的唯一标识符
    string SenderUUID = 5;
    // 指定接收者为指定类型的所有成员。如果指定了唯一接收者那么则忽略此选项。
    int32 Receiver = 6;
    // 指定唯一接收者
    string ReceiverUUID = 7;
    // 消息内容
    string Context = 8;
    // 扩展数据
    bytes ExtendData = 9;
    // 指示发送还是接收
    MsgTypes Type = 10;
}

message GLOBAL_CONFIG
{
    // Network 
    int32   ListenPort = 1;
    bool    EnableSSL=2;
    string  CertFilePath=3;
    string  KeyFilePath=4;
    string  CertPasswd=5;
    int32   ConnectTime=6;
    int32   ReceiveTime=7;

    // Log 
    string  LogPath = 8;
    int32   LogLevel =9 ;
    bool    DebugMode = 10;

    // Performance
    int32   MQThreadQuantity =11;// Message queue work thread quantity 
    int32   EPollThreadQuantity=12;// EPoll event process thread quantity
    int32   ProcessThreadQuantity=13;// work thread for DataPackage process quantity
    int32   PrioritySize=14;// Priority Size for the list. the priority level cannot bigger than this value.


    ModuleType Type = 15;
    bytes ExConfig = 16;
};

message CONTROL_CONFIG
{
    int32 BlockTime=1;
}

message DB_CONFIG
{
    string  ServerAddress=1;
    int32   ServerPort=2;
    string  User=3;
    string  Passwd=4;
    string  Database=5;
}

message DATA_CONFIG
{
    int32     DataReceivePort=1; 
    int32     DataRecvTime=2;
} 

message GATEWAY_CONFIG{
    int32     ClientCheckTime=1;
    string  ClientCheckKey=2;
    int32     TimeoutCheckInterval=3;
    int32       TimeoutTime=4;
    string      ProcessAddress=5;
};

message FIREWALL_CONFIG{
    int32 BlockTime=1;
}

message PROCESS_CONFIG{
    int32  LuaContextQuantity=1;
    string  LuaScriptFolder=2;
    string  LuaEntryFile=3;
    string  LuaEntryFunction=4;
    string  LuaProcessFunction=5;
    string  LuaSocketCloseFunction=6;
};
